### PR TITLE
[CI] Rework wheel packaging for compiler driver

### PR DESCRIPTION
**Pull request by rmoyard**
**Description of the Change:**

- Build compiler driver
- Remove unnecessary copy of files to the wheel
- Copy compiler-driver.so in the wheels
- Add `FromElementsOp(ir.OpView)` to extra bindings
- Fixes two memory issues in the C++ code base, unearthed by some tests
failing on the recent wheel build:
- A local variable was used outside of its stack scope. The variable was
captured by a lambda function whose lifetime exceeded that of the
variable.
- A ValueRange was generated without underlying storage that persists
for the lifetime of the range.

To prevent further such errors #49 should be resolved as soon as
possible.

Some other issues that were fixed:
- CMake was reconfigured such that building the dialects no longer
builds Enzyme, since the external project is built separately.
- Missing markers were added to new tensorflow tests. Tests should no
longer fail if tensorflow is not installed.
- Import Catalyst before TF

[sc-46569]

### CODE REVIEW

**Comment by erick-xanadu** [comment]
Yes, I think you got it by adding the target, but also the wheels will need to add the compiler_driver.so

So, what is missing is, somewhere in the catalyst/Makefile wheel recipe to copy over compiler_driver.so

**Comment by dime10** [comment]
> Yes, I think you got it by adding the target, but also the wheels will need to add the compiler_driver.so
> 
> So, what is missing is, somewhere in the catalyst/Makefile wheel recipe to copy over compiler_driver.so

The local recipe should work now after https://github.com/PennyLaneAI/catalyst/pull/289/commits/df1e082133702f1243eb89c3fa5a6f140330716d, I've also cleaned up to the wheel to only include the absolute minimum necessary.

Hopefully the CI script will succeed now.

**Review by erick-xanadu — APPROVED** [review]

**Comment by dime10** [comment]
Still waiting to see that the action completes successfully before merging this: https://github.com/PennyLaneAI/catalyst/actions/runs/6330671154/job/17193639275

File: .github/workflows/build-wheel-linux-x86_64.yaml [hunk 1]
```diff
@@ -260,9 +260,7 @@ jobs:
               -DPython3_EXECUTABLE=$(which python${{ matrix.python_version }}) \
               -DPython3_NumPy_INCLUDE_DIRS=$(python${{ matrix.python_version }} -c "import numpy as np; print(np.get_include())")
 
-        # TODO: when updating LLVM, test to see if mlir/unittests/Bytecode/BytecodeTest.cpp:55 is passing
-        # and remove filter
-        LIT_FILTER_OUT="Bytecode" cmake --build llvm-build --target check-mlir opt -j$(nproc)
+        cmake --build llvm-build --target check-mlir opt -j$(nproc)
```
Comment (rmoyard) [inline .github/workflows/build-wheel-linux-x86_64.yaml:263]:
```suggestion
        # TODO: when updating LLVM, test to see if mlir/unittests/Bytecode/BytecodeTest.cpp:55 is passing
        # and remove filter
        LIT_FILTER_OUT="Bytecode" cmake --build llvm-build --target check-mlir opt -j$(nproc)
```

**Comment by rmoyard** [comment]
Wheels: https://github.com/PennyLaneAI/catalyst/actions/runs/6339496113

**Comment by rmoyard** [comment]
Wheels are building but the tests are failing:
```
 FAILED frontend/test/pytest/test_autograph.py::TestSourceCodeInfo::test_non_converted_function - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestSourceCodeInfo::test_qjit - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestSourceCodeInfo::test_qnode - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestSourceCodeInfo::test_func - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestIntegration::test_unsupported_object - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestIntegration::test_lambda - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestIntegration::test_classical_function - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestIntegration::test_nested_function - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestIntegration::test_qnode - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestIntegration::test_indirect_qnode - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestIntegration::test_multiple_qnode - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestIntegration::test_nested_qnode - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestIntegration::test_nested_qjit - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestCodePrinting::test_unconverted - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestCodePrinting::test_lambda - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestCodePrinting::test_classical_function - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestCodePrinting::test_nested_function - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestCodePrinting::test_qnode - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestCodePrinting::test_indirect_qnode - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestCodePrinting::test_multiple_qnode - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestCodePrinting::test_nested_qnode - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestCodePrinting::test_nested_qjit - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestConditionals::test_simple_cond - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestConditionals::test_cond_one_else_if - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestConditionals::test_cond_many_else_if - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestConditionals::test_qubit_manipulation_cond[lightning.qubit] - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestConditionals::test_branch_return_mismatch[lightning.qubit] - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestConditionals::test_branch_multi_return_mismatch[lightning.qubit] - AssertionError: Regex pattern did not match.
 Regex: 'Conditional requires consistent return types across all branches'
 Input: "unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'"
FAILED frontend/test/pytest/test_autograph.py::TestForLoops::test_python_range_fallback - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestForLoops::test_for_in_array - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestForLoops::test_for_in_array_unpack - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestForLoops::test_for_in_numeric_list - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestForLoops::test_for_in_numeric_list_of_list - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestForLoops::test_for_in_object_list - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestForLoops::test_for_in_object_list_strict - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestForLoops::test_for_in_static_range - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestForLoops::test_for_in_static_range_indexing_numeric_list - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestForLoops::test_for_in_static_range_indexing_object_list - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestForLoops::test_for_in_dynamic_range - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestForLoops::test_for_in_dynamic_range_indexing_numeric_list - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestForLoops::test_for_in_dynamic_range_indexing_object_list - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestForLoops::test_for_in_enumerate_array - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestForLoops::test_for_in_enumerate_array_no_unpack - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestForLoops::test_for_in_enumerate_nested_unpack - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestForLoops::test_for_in_enumerate_start - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestForLoops::test_for_in_enumerate_numeric_list - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestForLoops::test_for_in_enumerate_object_list - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestForLoops::test_for_in_other_iterable_object - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestForLoops::test_loop_carried_value - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestForLoops::test_iteration_element_access - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestForLoops::test_temporary_loop_variable - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestForLoops::test_uninitialized_variables - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestForLoops::test_init_with_invalid_jax_type - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestForLoops::test_init_with_mismatched_type - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestForLoops::test_ignore_warnings - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestMixed::test_no_python_loops - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_autograph.py::TestMixed::test_cond_if_for_loop_for - TypeError: unsupported operand type(s) for |: '_CallableGenericAlias' and 'NoneType'
FAILED frontend/test/pytest/test_conditionals.py::TestCond::test_qubit_manipulation_cond[lightning.qubit]
FAILED frontend/test/pytest/test_jax_integration.py::TestJAXJIT::test_multiple_calls[lightning.qubit]
FAILED frontend/test/pytest/test_mid_circuit_measurement.py::TestMidCircuitMeasurement::test_basic[lightning.qubit] - RuntimeError: random_device could not be read
FAILED frontend/test/pytest/test_mid_circuit_measurement.py::TestMidCircuitMeasurement::test_more_complex[lightning.qubit]
FAILED frontend/test/pytest/test_jit_behaviour.py::TestCallsiteCompileVsFunctionDefinitionCompile::test_equivalence[lightning.qubit] - RuntimeError: random_device could not be read
FAILED frontend/test/pytest/test_jit_behaviour.py::TestDecorator::test_function_is_cached[lightning.qubit] - RuntimeError: random_device could not be read
FAILED frontend/test/pytest/test_jit_behaviour.py::TestCaching::test_function_is_cached[lightning.qubit] - RuntimeError: random_device could not be read
FAILED frontend/test/pytest/test_jit_behaviour.py::TestShots::test_shots_in_decorator_in_sample[lightning.qubit]
FAILED frontend/test/pytest/test_measurements_results.py::TestOtherMeasurements::test_multiple_return_values[lightning.qubit]
FAILED frontend/test/pytest/test_pytree_args.py::TestPyTreesFuncArgs::test_args_used_in_measure[lightning.qubit]
FAILED frontend/test/pytest/test_loops.py::TestWhileLoops::test_alternating_loop[lightning.qubit]
FAILED frontend/test/pytest/test_variable_wires.py::TestBasicCircuits::test_measurements[lightning.qubit-args0]
FAILED frontend/test/pytest/test_pytree_args.py::TestPyTreesReturnValues::test_return_value_float[lightning.qubit] - RuntimeError: random_device could not be read
FAILED frontend/test/pytest/test_pytree_args.py::TestPyTreesReturnValues::test_return_value_tuples[lightning.qubit]
FAILED frontend/test/pytest/test_loops.py::TestWhileLoops::test_closure_condition_fn[lightning.qubit]
```

**Comment by erick-xanadu** [comment]
@rmoyard let me try to build the wheel locally and test it that way, that would at least give us a hint at what the problem might be.

**Comment by rmoyard** [comment]
To be fixed:
```
FAILED frontend/test/pytest/test_autograph.py::TestConditionals::test_qubit_manipulation_cond[lightning.qubit] - RuntimeError: random_device could not be read
FAILED frontend/test/pytest/test_conditionals.py::TestCond::test_qubit_manipulation_cond[lightning.qubit]
FAILED frontend/test/pytest/test_jax_integration.py::TestJAXJIT::test_multiple_calls[lightning.qubit]
FAILED frontend/test/pytest/test_mid_circuit_measurement.py::TestMidCircuitMeasurement::test_basic[lightning.qubit] - RuntimeError: random_device could not be read
FAILED frontend/test/pytest/test_mid_circuit_measurement.py::TestMidCircuitMeasurement::test_more_complex[lightning.qubit] - RuntimeError: random_device could not be read
FAILED frontend/test/pytest/test_pytree_args.py::TestPyTreesReturnValues::test_return_value_float[lightning.qubit] - RuntimeError: random_device could not be read
FAILED frontend/test/pytest/test_pytree_args.py::TestPyTreesReturnValues::test_return_value_tuples[lightning.qubit]
FAILED frontend/test/pytest/test_measurements_results.py::TestOtherMeasurements::test_multiple_return_values[lightning.qubit]
FAILED frontend/test/pytest/test_jit_behaviour.py::TestCallsiteCompileVsFunctionDefinitionCompile::test_equivalence[lightning.qubit] - RuntimeError: random_device could not be read
FAILED frontend/test/pytest/test_jit_behaviour.py::TestDecorator::test_function_is_cached[lightning.qubit] - RuntimeError: random_device could not be read
FAILED frontend/test/pytest/test_pytree_args.py::TestPyTreesFuncArgs::test_args_used_in_measure[lightning.qubit] - RuntimeError: random_device could not be read
FAILED frontend/test/pytest/test_jit_behaviour.py::TestCaching::test_function_is_cached[lightning.qubit] - RuntimeError: random_device could not be read
FAILED frontend/test/pytest/test_pytree_args.py::TestPyTreesFuncArgs::test_args_used_indirectly_in_measure[lightning.qubit] - RuntimeError: random_device could not be read
FAILED frontend/test/pytest/test_jit_behaviour.py::TestShots::test_shots_in_decorator_in_sample[lightning.qubit] - RuntimeError: random_device could not be read
FAILED frontend/test/pytest/test_jit_behaviour.py::TestShots::test_shots_in_callsite_in_sample[lightning.qubit]
FAILED frontend/test/pytest/test_qnode.py::test_variable_capture[0-False]
FAILED frontend/test/pytest/test_pytree_args.py::TestPyTreesReturnValues::test_return_value_dict[lightning.qubit]
FAILED frontend/test/pytest/test_loops.py::TestWhileLoops::test_alternating_loop[lightning.qubit]
FAILED frontend/test/pytest/test_conditionals.py::TestCond::test_identical_branch_names[lightning.qubit] - RuntimeError: random_device could not be read
FAILED frontend/test/pytest/test_contexts.py::TestTracing::test_fixed_tracing[lightning.qubit] - RuntimeError: random_device could not be read
FAILED frontend/test/pytest/test_contexts.py::TestTracing::test_cond_inside_while_loop[lightning.qubit]
FAILED frontend/test/pytest/test_variable_wires.py::TestBasicCircuits::test_measurements[lightning.qubit-args0]
```

**Comment by rmoyard** [comment]
https://github.com/PennyLaneAI/catalyst/actions/runs/6396499235 :crossed_fingers:

**Review by erick-xanadu — APPROVED** [review]
:)

**Comment by josh146** [comment]
🎉

**Pull request merged**